engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TypeCheckValueNode.java

```diff
@@ -164,8 +178,9 @@ public static TypeCheckValueNode fail(String comment) {
    * @return node performing the check
    */
   public static TypeCheckValueNode single(String comment, Type expectedType) {
-    var typeCheckNodeImpl = SingleTypeCheckNodeGen.create(comment, expectedType);
-    return new TypeCheckValueNode(typeCheckNodeImpl, true);
+    var checkNode = SingleTypeCheckNodeGen.create(comment, expectedType);
+    var allTypes = comment == null || !comment.startsWith("the result of `");
```

Comment:
- pretty ugly check introduced by 968ea7d
- but it fixes `Extract.B.hidden.type` test case
- alternative fixes:
   - duplicate `Type.Ascription` IR
   - instead of comment hold "purpose" hold more structured info to know whether consider `allTypes` or not
- in any case that will invalidate caches

Reply:
Resolved by introducing `AscriptionReason` in 54a5c2e082f8a72c5f4c69af9f526251e0095168